client-tools/Shell/ProcessMonitoringFeature.cpp

```diff
@@ -0,0 +1,158 @@
+///
+/// @author Wilfried Goesgens
+////////////////////////////////////////////////////////////////////////////////
+
+#include <chrono>
+
+#include "Basics/Mutex.h"
+#include "Basics/MutexLocker.h"
+#include "Basics/application-exit.h"
+
+#include "Logger/LogMacros.h"
+#include "Logger/Logger.h"
+#include "Logger/LoggerStream.h"
+
+#include "ApplicationFeatures/ApplicationServer.h"
+
+#include "V8/v8-deadline.h"
+#include "V8/V8SecurityFeature.h"
+
+#include "ProcessMonitoringFeature.h"
+
+using namespace arangodb::basics;
+using namespace arangodb::options;
+using namespace arangodb;
+
+namespace arangodb {
+
+void ProcessMonitoringFeature::addMonitorPID(ExternalId const& pid) {
+  MUTEX_LOCKER(mutexLocker, _MonitoredExternalProcessesLock);
+  _monitoredProcesses.push_back(pid);
+}
+
+void ProcessMonitoringFeature::removeMonitorPID(ExternalId const& pid) {
+  {
+    MUTEX_LOCKER(mutexLocker, _MonitoredExternalProcessesLock);
+    for (auto it = _monitoredProcesses.begin(); it != _monitoredProcesses.end();
+         ++it) {
+      if (it->_pid == pid._pid) {
+        _monitoredProcesses.erase(it);
+        break;
+      }
+    }
+  }
+  // make sure its really not monitored anymore once we exit:
+  std::this_thread::sleep_for(std::chrono::milliseconds(110));
+}
+
+std::optional<ExternalProcessStatus>
+ProcessMonitoringFeature::getHistoricStatus(TRI_pid_t pid) {
+  MUTEX_LOCKER(mutexLocker, _MonitoredExternalProcessesLock);
+  auto it = _ExitedExternalProcessStatus.find(pid);
+  if (it == _ExitedExternalProcessStatus.end()) {
+    return std::nullopt;
+  }
+  return std::optional<ExternalProcessStatus>{it->second};
+}
+
+ProcessMonitoringFeature::ProcessMonitoringFeature(Server& server)
+    : ArangoshFeature{server, *this} {
+  startsAfter<V8SecurityFeature>();
+  _monitoredProcesses.reserve(10);
+}
+
+ProcessMonitoringFeature::~ProcessMonitoringFeature() = default;
+
+void ProcessMonitoringFeature::validateOptions(
+    std::shared_ptr<options::ProgramOptions>) {
+  _enabled =
+      server().getFeature<V8SecurityFeature>().isAllowedToControlProcesses();
+}
+
+void ProcessMonitoringFeature::start() {
+  if (_enabled) {
+    _monitorThread = std::make_unique<ProcessMonitorThread>(server(), *this);
+    if (!_monitorThread->start()) {
+      LOG_TOPIC("33333", FATAL, Logger::SYSCALL)
+          << "failed to launch monitoring background thread";
+      FATAL_ERROR_EXIT();
+    }
+  }
+}
+void ProcessMonitoringFeature::beginShutdown() {
+  if (_enabled) {
+    _monitorThread->shutdown();
+  }
+}
+
+void ProcessMonitoringFeature::stop() {
+  if (_enabled) {
+    _monitorThread->shutdown();
+  }
+}
+
+void ProcessMonitorThread::run() {  // override
+  std::vector<ExternalId> mp;
+  mp.reserve(10);
+  while (!isStopping()) {
+    try {
+      {
+        MUTEX_LOCKER(mutexLocker,
+                     _processMonitorFeature._MonitoredExternalProcessesLock);
+        mp = _processMonitorFeature._monitoredProcesses;
```

Comment:
you probably want swap here and clear after for.
Otherwise reserve is completely useless (we just alloc and dealloc this memory)